packages/gatsby-plugin-fastify/src/plugins/headers.ts

```diff
@@ -0,0 +1,44 @@
+import fp from "fastify-plugin";
+import { isMatch } from "picomatch";
+import type { FastifyPluginAsync, FastifyReply } from "fastify";
+import { getConfig } from "../utils/config";
+import { NEVER_CACHE_HEADER } from "../utils/constants";
+
+export const handleHeaders: FastifyPluginAsync = fp(async (fastify) => {
+  fastify.addHook("onSend", async function (request, reply: FastifyReply) {
+    // add all necessary headers
+    const {
+      server: {
+        headers,
+        features: {
+          headers: { useDefaultCaching },
+        },
+      },
+    } = getConfig();
+
+    // get the url without query params
+    const url = request.raw.url?.split(`?`)[0];
+
+    // handle directory URLs which resolve to index.html
+    if (
+      useDefaultCaching &&
+      (url?.endsWith("/") || // directories with trailing slash
+        (!url?.includes(".") && !url?.endsWith("/"))) // no trailing slash or file extension, so it's a directory
+    ) {
+      reply.headers(NEVER_CACHE_HEADER);
+    }
```

Comment:
>What would would this NOT apply to? 

It shouldn't apply to any specific static files (images, js, css, pdfs, docs, etc) just `directories` either with/without a trailing slash, where the actual served file is the `/index.html` so it should only apply `NEVER_CACHE_HEADER` to `.html` files (and function routes: see below) which matches gatsby docs/hosting. It could be missing something, but I think it's right. 

> I'm also wondering if it is still to broad... this shouldn't apply to functions

You are correct in that this will apply to function routes like `/api/test/`. I've updated the `plugin-fastify` test site to support running functions on gatsby hosting in 9a328fb662ff04347d63d235798f3bef1b538eb4 and verified that gatsby hosting [does indeed](https://pluginsfeatfastifydefaultcusto.gatsbyjs.io/api/test) use the `NEVER_CACHE_HEADER` for function routes as well. 

![image](https://user-images.githubusercontent.com/4126987/225116459-96e0b600-307a-4edb-9318-0e70b2cd7268.png)

> Maybe we need to tag routes with their type e.g. like the way x-gatsby-server does but structure the data so that its SSR|| DSG || STATIC || FUNCTIONS etc...

In the first refactor I did go down the path of tagging routes with their module by adding a decorator `reply.module` which I was going to then use in the headers plugin to set the `FG_MODULE_HEADER` based on the module used as well. I had this working for all modules except `static.ts` because it is globally scoped, since the `fastify-static` plugin doesn't have any hooks to run when it's actually serving a file the `Static` module was added to all replies, not just the ones actually served by static.

I tried quite a few methods such as narrowing the scope of static (however 500/404/client-routes needed `reply.sendFile`), adding separate `fastify-static` implementations within 500/404/client, adding the `reply.module` within `fastify-static`'s `setHeaders` option, and some others I can't recall now...I was not able to reliably only add the static module to files actually served by `fastify-static` though I may have been missing something obvious? 

> Also, where is the immutable cache header getting applied?

See dupe question below which is more specific: https://github.com/gatsby-uc/plugins/pull/392#discussion_r1134965306

Reply:
resolved in 897e7d519c1752f8ce150f1d583dcd8728b98ab2